### ISSUE TITLE
Single OCI registry: Stage 1 - container images

### ISSUE THREAD

**Issue by piontec** [giantswarm/roadmap#2882]
## Related epics
- Stage 2: #3056
- Retagger changes: https://github.com/giantswarm/roadmap/issues/3068

## Tasks
We discussed the new registry architecture - [RFC](https://github.com/giantswarm/rfc/tree/main/single-registry-architecture).

We have to get out of Docker Hub by the end of 2023 - this is called 'stage 1' in the RFC. This ticket tracks and groups tasks around that.

```[tasklist]
### Step 0 - Preparations
- [x] [Planning and RFC related issue](https://github.com/giantswarm/giantswarm/issues/27243)
- [ ] https://github.com/giantswarm/roadmap/issues/2896
```

```[tasklist]
### Step 1 - Setting up the new registry
- [ ] https://github.com/giantswarm/roadmap/issues/2895
- [ ] https://github.com/giantswarm/roadmap/issues/2915
- [ ] https://github.com/giantswarm/roadmap/issues/2916
- [ ] https://github.com/giantswarm/roadmap/issues/2926
- [ ] https://github.com/giantswarm/roadmap/issues/3025
- [ ] https://github.com/giantswarm/roadmap/issues/3027
```

```[tasklist]
### Step 1.5 - Handling private images
- [ ] https://github.com/giantswarm/roadmap/issues/3026
- [ ] https://github.com/giantswarm/roadmap/issues/3020
- [ ] https://github.com/giantswarm/roadmap/issues/3064
- [ ] https://github.com/giantswarm/roadmap/issues/3063
- [ ] https://github.com/giantswarm/roadmap/issues/3062
- [ ] https://github.com/giantswarm/roadmap/issues/3041
- [ ] https://github.com/giantswarm/roadmap/issues/3084
```

```[tasklist]
### Step 2 - Phase out Docker Hub usage
- [ ] https://github.com/giantswarm/giantswarm/issues/28933
- [ ] https://github.com/giantswarm/giantswarm/issues/30502
- [ ] https://github.com/giantswarm/roadmap/issues/3409
- [ ] https://github.com/giantswarm/roadmap/issues/3179
- [ ] https://github.com/giantswarm/roadmap/issues/3080
- [ ] https://github.com/giantswarm/roadmap/issues/3112
- [ ] https://github.com/giantswarm/roadmap/issues/3110
- [ ] https://github.com/giantswarm/roadmap/issues/3410
- [ ] https://github.com/giantswarm/roadmap/issues/2971
- [ ] https://github.com/giantswarm/roadmap/issues/2979
- [ ] https://github.com/giantswarm/roadmap/issues/3108
- [ ] https://github.com/giantswarm/roadmap/issues/3031
- [ ] https://github.com/giantswarm/roadmap/issues/3032
- [ ] https://github.com/giantswarm/roadmap/issues/3028
- [ ] https://github.com/giantswarm/roadmap/issues/3017
- [ ] https://github.com/giantswarm/roadmap/issues/3029
- [ ] https://github.com/giantswarm/roadmap/issues/3030
- [x] Figure out what to do with private images that we have on quay.io
- [ ] https://github.com/giantswarm/roadmap/issues/3044
- [ ] https://github.com/giantswarm/roadmap/issues/3109
- [ ] https://github.com/giantswarm/roadmap/issues/3176
- [ ] https://github.com/giantswarm/roadmap/issues/3111
- [ ] https://github.com/giantswarm/roadmap/issues/3177
- [ ] https://github.com/giantswarm/roadmap/issues/3168
- [ ] https://github.com/giantswarm/roadmap/issues/3169
- [ ] https://github.com/giantswarm/roadmap/issues/3244
- [ ] https://github.com/giantswarm/roadmap/issues/3245
- [ ] https://github.com/giantswarm/roadmap/issues/3246
- [ ] https://github.com/giantswarm/roadmap/issues/3247
- [ ] https://github.com/giantswarm/roadmap/issues/3248
```

```[tasklist]
### Step 3 - wrapping up
- [ ] cleanup old registries when they are not used anymore
- [ ] update docs about registries
- [ ] remove pushes to old registries from retagger
```

```[tasklist]
### Security enhancements
- [ ] use `cosign` to sign images built on circle [blocked by circleci's support for cosign]
- [ ] generate SBoMs automatically for all image builds and upload results to the registry
```

**Comment by marians** [comment]
We should announce to customers in time that this change is going to happen. Here are the details we should include IMO.

- Our docker.io credentials are going to be removed from the hosts (date to be discussed)
- When using images from docker.io in workloads, please provide your own ImagePullSecret. You can start doing this immediately, no need to wait.
- Otherwise you might get affected by (probably) IP-based rate limiting. Actually Docker does not disclose details. By using your own credentials via ImagePullSecrets, you make sure that pulls are accounted for via your account(s) and you have more control over this.
- Side note: most customers use their own registries and copy the images they need from Docker Hub over to their own registries. This is the method we would recommend.

Please let's verify this messaging and also fill in the missing date info.

**Comment by marians** [comment]
📉 Metrics on containers using docker.io images are now available in [this Grafana Cloud dashboard](https://giantswarm.grafana.net/d/ca1f873e-8822-482f-9a06-1e715ab0f99e/container-images-from-docker-hub?orgId=1)

**Comment by piontec** [comment]
Ad Step 2. Notes from Whites:

MCs:
- how base images registry domains are set up? are they configurable?
for systemd images we use a configurable docker registry. Default is [docker.io](http://docker.io/) and we can configure it in `installations` repo.
Example: https://github.com/giantswarm/installations/blob/d6a674395835a4b06e37126f485b4dc1cf94b1ad/akita/terraform/bootstrap.sh#L69
Changing needs rolling of all MC nodes.

WCs:
- how base images registry domains are set up? are they configurable?
This is controller by aws-operator and configurable via config repo.
Default is [docker.io](http://docker.io/) (https://github.com/giantswarm/config/blob/main/default/config.yaml#L10)
Can be overridden per installation (https://github.com/giantswarm/config/blob/main/installations/akita/config.yaml.patch#L17)
- how do apps get registry domain set up?
this one I'm not 100% sure, but I think this is all controlled by app operator (https://github.com/giantswarm/config/blob/main/default/apps/app-operator/configmap-values.yaml.template#L2) and it should be identical to how we did for aws-operator in the previous point.
- how do we set up registry credentials?
We set credentials as part of containerd config file (https://github.com/giantswarm/k8scloudconfig/blob/master/files/conf/containerd-config.toml#L45) and docker config file (https://github.com/giantswarm/k8scloudconfig/blob/5ddefddd6650b16ffd0f4388e95ec7e7c12aff0d/files/conf/kubelet-docker-config.json#L4).
Value is injected by aws-operator and I believe is configured by MC in the config repo (example https://github.com/giantswarm/config/blob/6dea50d889fe8af4769d151d359d1e2ca10e27bc/installations/visitor/secret.yaml#L13)
Metrics:
- check if we have containerd metrics about pulls and since which release
I think this should be included in containerd 1.7.0.
Not shipped with flatcar stable yet :(

**Comment by marians** [comment]
I looked for internal documentation regarding registries, to be updated/reviewed. So far I found

- [Registry mirrors](https://handbook.giantswarm.io/docs/product/architecture-specs-adrs/specs/registry-mirror/)
- [Creating container registries for retagger](https://intranet.giantswarm.io/docs/dev-and-releng/container-registry/)
- [Changes in tooling to handle multiple image registries
](https://intranet.giantswarm.io/docs/product/architecture-specs-adrs/specs/multiple-image-registries/)

**Comment by marians** [comment]
Since it has gotten tedious to track PRs for more than 150 repos, I created this spreadsheet:

https://docs.google.com/spreadsheets/d/1CwK4a9Q9AKZq_n2rYNcMBfUaXAaxFXCm_gvzhMmjgCM/edit#gid=0

The idea is:

1. CircleCI PR should be merged first. Extra issue: https://github.com/giantswarm/roadmap/issues/2979
2. Then default domain can be adjusted via another PR. (Not all of the repos we treated in (1) actually have a chart or a default domain setting. In this case it is probably not a deployed workload that generates many pulls, so these aren't important.) - Extra issue: https://github.com/giantswarm/roadmap/issues/3017
3. Then a release is made

**Comment by alex-dabija** [comment]
I tested the creation of a CAPA cluster with the new registry (`gsoci.azurecr.io`) with the following configuration:
```
global:
  components:
    containerd:
      containerRegistries:
        docker.io:
          - endpoint: gsoci.azurecr.io
        gsoci.azurecr.io:
          - endpoint: gsoci.azurecr.io
  connectivity:
    availabilityZoneUsageLimit: 3
    network: {}
    topology: {}
  controlPlane: {}
  metadata:
    name: alex02
    organization: giantswarm
  nodePools:
    nodepool0:
      instanceType: m5.xlarge
      maxSize: 3
      minSize: 3
      rootVolumeSizeGB: 300
  providerSpecific: {}
```

I couldn't remove `docker.io` entirely because of how Helm merges dicts.

The cluster had the following `containerd` config:
```
version = 2

# recommended defaults from https://github.com/containerd/containerd/blob/main/docs/ops.md#base-configuration
# set containerd as a subreaper on linux when it is not running as PID 1
subreaper = true
# set containerd's OOM score
oom_score = -999
disabled_plugins = []
[plugins."io.containerd.runtime.v1.linux"]
# shim binary name/path
shim = "containerd-shim"
# runtime binary name/path
runtime = "runc"
# do not use a shim when starting containers, saves on memory but
# live restore is not supported
no_shim = false

[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
# setting runc.options unsets parent settings
runtime_type = "io.containerd.runc.v2"
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
SystemdCgroup = true
[plugins."io.containerd.grpc.v1.cri"]
sandbox_image = "quay.io/giantswarm/pause:3.9"

[plugins."io.containerd.grpc.v1.cri".registry]
  [plugins."io.containerd.grpc.v1.cri".registry.mirrors]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
      endpoint = ["https://gsoci.azurecr.io",]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."gsoci.azurecr.io"]
      endpoint = ["https://gsoci.azurecr.io",]
[plugins."io.containerd.grpc.v1.cri".registry.configs]
```

The cluster had the following images running:
```
kubectl --context gs-grizzly-alex02-clientcert get pods --all-namespaces -o jsonpath="{.items[*].spec['initContainers', 'containers'][*].image}" |\
tr -s '[[:space:]]' '\n' |\
sort |\
uniq -c
      4 docker.io/giantswarm/aws-ebs-csi-driver:v1.21.0
      3 docker.io/giantswarm/aws-ebs-csi-volume-limiter:0.1.0
      7 docker.io/giantswarm/cert-exporter:2.8.4
      1 docker.io/giantswarm/cert-manager-cainjector:v1.12.4
      2 docker.io/giantswarm/cert-manager-webhook:v1.12.4
      2 docker.io/giantswarm/cilium-operator-generic:v1.14.3
     42 docker.io/giantswarm/cilium:v1.14.3
      3 docker.io/giantswarm/configmap-reload:v0.8.0
      3 docker.io/giantswarm/coredns:1.11.1
      1 docker.io/giantswarm/csi-attacher:v4.4.1
      3 docker.io/giantswarm/csi-node-driver-registrar:v2.9.0
      1 docker.io/giantswarm/csi-provisioner:v3.6.1
      1 docker.io/giantswarm/csi-resizer:v1.8.1
      3 docker.io/giantswarm/etcd:3.5.6-0
      1 docker.io/giantswarm/etcd-kubernetes-resources-count-exporter:1.8.0
      1 docker.io/giantswarm/external-dns:v0.11.0
      3 docker.io/giantswarm/grafana-agent:v0.37.2
      1 docker.io/giantswarm/hubble-relay:v1.14.3
      1 docker.io/giantswarm/hubble-ui-backend:v0.12.1
      1 docker.io/giantswarm/hubble-ui:v0.12.1
      3 docker.io/giantswarm/kube-apiserver:v1.24.16
      3 docker.io/giantswarm/kube-controller-manager:v1.24.16
      3 docker.io/giantswarm/kube-scheduler:v1.24.16
      1 docker.io/giantswarm/kube-state-metrics:v2.10.0
      4 docker.io/giantswarm/livenessprobe:v2.11.0
      6 docker.io/giantswarm/node-exporter:v1.3.1
      2 docker.io/giantswarm/prometheus-config-reloader:v0.68.0
      1 docker.io/giantswarm/prometheus-operator:v0.68.0
      1 docker.io/giantswarm/prometheus:v2.47.1
      6 docker.io/giantswarm/promtail:2.8.4
      2 docker.io/giantswarm/vpa-admission-controller:0.14.0
      1 docker.io/giantswarm/vpa-recommender:0.14.0
      1 docker.io/giantswarm/vpa-updater:0.14.0
      1 public.ecr.aws/gravitational/teleport-distroless:14.1.3
      3 quay.io/giantswarm/amazon-eks-pod-identity-webhook-gs:v0.2.0
      3 quay.io/giantswarm/aws-cloud-controller-manager:v1.24.1
      6 quay.io/giantswarm/capi-node-labeler-app:0.3.4
      1 quay.io/giantswarm/cert-manager-controller:v1.12.4
      2 quay.io/giantswarm/chart-operator:2.35.2
      1 quay.io/giantswarm/cluster-autoscaler:v1.27.3
      6 quay.io/giantswarm/docker-kubectl:1.25.4
      2 quay.io/giantswarm/metrics-server:v0.6.4
      6 quay.io/giantswarm/net-exporter:1.18.0
```

I'm assuming that the `docker.io` images are pulled from `gsoci.azurecr.io` because it's the only endpoint configured.

```[tasklist]
## TODOs for KaaS:
- [ ] Make the `gsoci.azurecr.io` registry the default;
- [ ] Update `imageRepository: docker.io/giantswarm` for the kubeadm control plane configuration in the cluster-aws chart;
- [ ] Update `imageRepository: docker.io/giantswarm` for the kubeadm control plane configuration in the cluster-aws chart;
- [ ] Make `public.ecr.aws/gravitational/teleport-distroless:14.1.3` available from our registry;
- [ ] Replace `quay.io` with the new registry.

**Comment by weatherhog** [comment]
@piontec what is missing here for stage 1. Can this be closed?

**Comment by piontec** [comment]
Stage 1 is complete, 1.5 as well, but this is a big epic that covers next steps, that are not done yet. For me it's a great overview of where we are in this big story and I would keep this open, even if we're not currently actively working on it.
